File: src/Billing/Mediation/Summary/Cache/ICacheble.cs
// Ported from legacy Models_Mediation/EntityExtensions/ICachebale.cs + Mediation/Cache/{CachedItem,
// CachedSummary}.cs. The cache contract (per-row SQL builders), the cached-item wrapper, the merge type,
// and the new-id source. Trimmed to the members the summary path uses.
#nullable disable
using System;
using System.Text;

namespace MediationModel
{
    public interface ICacheble { }

    /// <summary>An entity the cache can persist: it builds its own INSERT-values / UPDATE / DELETE SQL
    /// fragments (legacy ICacheble&lt;TEntity&gt;).</summary>
    public interface ICacheble<TEntity> : ICacheble
    {
        StringBuilder GetExtInsertValues();
        StringBuilder GetUpdateCommand(Func<TEntity, string> whereClauseMethod);
        StringBuilder GetDeleteCommand(Func<TEntity, string> whereClauseMethod);
    }

    /// <summary>Add (new CDRs) or Substract (erased/reprocessed CDRs) when merging a summary (legacy spelling kept).</summary>
    public enum SummaryMergeType { Add, Substract }

    /// <summary>What the cache returns for a keyed item (legacy CachedItem).</summary>
    public sealed class CachedItem<TKey, TEntity>
    {
        public TKey Key { get; }
        public TEntity Entity { get; }
        public CachedItem(TKey key, TEntity entity) { Key = key; Entity = entity; }
    }

    /// <summary>Source of new row ids per entity/table (legacy IAutoIncrementManager; keyed by table name
    /// here instead of the AutoIncrementCounterType enum).</summary>
    public interface IAutoIncrementManager
    {
        long GetNewCounter(string entityOrTableName);
    }
}

File: java/src/main/java/com/telcobright/billing/mediation/sql/ICacheble.java
// Ported from legacy Models_Mediation/EntityExtensions/ICachebale.cs + Mediation/Cache/{CachedItem,
// CachedSummary}.cs. The cache contract (per-row SQL builders). Trimmed to the members the summary path
// uses.
//
// NOTE (faithful-port deviation): the C# file declared BOTH a non-generic marker `interface ICacheble {}`
// AND `interface ICacheble<TEntity> : ICacheble`. Java cannot have two top-level types with the same name
// (the non-generic marker would be the raw type of the generic one), so the marker is subsumed into the
// raw form of this single generic interface. The other top-level types from the C# file
// (CachedItem, SummaryMergeType, IAutoIncrementManager) are split into their own files in this package.
package com.telcobright.billing.mediation.sql;

import java.util.function.Function;

/**
 * An entity the cache can persist: it builds its own INSERT-values / UPDATE / DELETE SQL fragments
 * (legacy ICacheble&lt;TEntity&gt;).
 */
public interface ICacheble<TEntity> {
    StringBuilder GetExtInsertValues();
    StringBuilder GetUpdateCommand(Function<TEntity, String> whereClauseMethod);
    StringBuilder GetDeleteCommand(Function<TEntity, String> whereClauseMethod);
}
